transport/handler_server.go

```diff
@@ -319,7 +320,16 @@ func (ht *serverHandlerTransport) HandleStreams(startStream func(*Stream)) {
 				s.buf.put(&recvMsg{data: buf[:n]})
 			}
 			if err != nil {
-				s.buf.put(&recvMsg{err: err})
+				if err, ok := err.(http2.StreamError); ok {
+					code := http2ErrConvTab[err.Code]
+					if code == codes.OK {
+						s.buf.put(&recvMsg{err: io.EOF})
+					} else {
+						s.buf.put(&recvMsg{err: StreamErrorf(code, "%v", err)})
+					}
+				} else {
+					s.buf.put(&recvMsg{err: io.EOF})
```

Comment:
BTW, the current change makes the test passed (so that you cannot reproduce) but again I am not sure my error handling code is correct (e.g., could Body.Read returns a http2.ConnectionError besides http2.StreamError).